br/pkg/lightning/backend/local/region_job.go

```diff
@@ -0,0 +1,620 @@
+	splitCli split.SplitClient,
+	supportMultiIngest bool,
+	checkWriteStall bool,
+) error {
+	switch j.stage {
+	case regionScanned, ingested:
+		return nil
+	case wrote:
+	}
+
+	if len(j.writeResult.sstMeta) == 0 {
+		j.convertStageTo(ingested)
+		return nil
+	}
+
+	for retry := 0; retry < maxRetryTimes; retry++ {
+		resp, err := j.doIngest(ctx, clientFactory, supportMultiIngest, checkWriteStall)
+		if err == nil && resp.GetError() == nil {
+			j.convertStageTo(ingested)
+			return nil
+		}
+		if err != nil {
+			if common.IsContextCanceledError(err) {
+				return err
+			}
+			log.FromContext(ctx).Warn("meet underlying error, will retry ingest",
+				log.ShortError(err), logutil.SSTMetas(j.writeResult.sstMeta),
+				logutil.Region(j.region.Region), logutil.Leader(j.region.Leader))
+			continue
+		}
+		canContinue, err := j.fixIngestError(ctx, resp, splitCli)
+		if common.IsContextCanceledError(err) {
+			return err
+		}
+		if !canContinue {
+			log.FromContext(ctx).Warn("meet error and handle the job later",
+				zap.Stringer("job stage", j.stage),
+				logutil.ShortError(j.lastRetryableErr),
+				logutil.Region(j.region.Region),
+				logutil.Key("start", j.keyRange.start),
+				logutil.Key("end", j.keyRange.end))
+			return nil
+		}
+		log.FromContext(ctx).Warn("meet error and will doIngest region, again",
+			logutil.ShortError(j.lastRetryableErr),
+			logutil.Region(j.region.Region),
+			logutil.Key("start", j.keyRange.start),
+			logutil.Key("end", j.keyRange.end))
+	}
+	return nil
+}
+
+func (j *regionJob) checkWriteStall(
+	ctx context.Context,
+	region *split.RegionInfo,
+	clientFactory ImportClientFactory,
+) (bool, *sst.IngestResponse, error) {
+	for _, peer := range region.Region.GetPeers() {
+		cli, err := clientFactory.Create(ctx, peer.StoreId)
+		if err != nil {
+			return false, nil, errors.Trace(err)
+		}
+		// currently we use empty MultiIngestRequest to check if TiKV is busy.
+		// If in future the rate limit feature contains more metrics we can switch to use it.
+		resp, err := cli.MultiIngest(ctx, &sst.MultiIngestRequest{})
+		if err != nil {
+			return false, nil, errors.Trace(err)
+		}
+		if resp.Error != nil && resp.Error.ServerIsBusy != nil {
+			return true, resp, nil
+		}
+	}
+	return false, nil, nil
+}
+
+// doIngest send ingest commands to TiKV based on regionJob.writeResult.sstMeta.
+// When meet error, it will remove finished sstMetas before return.
+func (j *regionJob) doIngest(
+	ctx context.Context,
+	clientFactory ImportClientFactory,
+	supportMultiIngest bool,
+	checkWriteStall bool,
+) (*sst.IngestResponse, error) {
+	if checkWriteStall {
+		writeStall, resp, err := j.checkWriteStall(ctx, j.region, clientFactory)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if writeStall {
+			return resp, nil
+		}
+	}
+
+	batch := 1
+	if supportMultiIngest {
+		batch = len(j.writeResult.sstMeta)
+	}
+
+	var resp *sst.IngestResponse
+	for i := 0; i < len(j.writeResult.sstMeta); i += batch {
```

Comment:
`i` can be renamed to `start`

Reply:
I suspect there's a bug because we use `batch` twice. Will check it later

```
	for i := 0; i < len(j.writeResult.sstMeta); i += batch {
		start := i * batch
		end := mathutil.Min((i+1)*batch, len(j.writeResult.sstMeta))
```

Reply:
https://github.com/pingcap/tidb/pull/40692/commits/5235648ce6c8866758f77ff7b769a95ac2a66d82